From 5383b510554d7ca12bacbc019f74fa69af79004e Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:54:40 -0400
Subject: [fs] jbd2: add transaction to checkpoint list earlier

Message-id: <1414058125-4183-127-git-send-email-lczerner@redhat.com>
Patchwork-id: 98207
O-Subject: [RHEL 7.1 PATCH v2 126/171] jbd2: add transaction to checkpoint list earlier
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Theodore Ts'o <tytso@mit.edu>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit d4e839d4a9dc31d0c229e616146b01e1ace56604

 We don't otherwise need j_list_lock during the rest of commit phase
 #7, so add the transaction to the checkpoint list at the very end of
 commit phase #6.  This allows us to drop j_list_lock earlier, which is
 a good thing since it is a super hot lock.

 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/jbd2/commit.c b/fs/jbd2/commit.c
index af36252..5f26139 100644
--- a/fs/jbd2/commit.c
+++ b/fs/jbd2/commit.c
@@ -1065,6 +1065,25 @@ restart_loop:
   goto restart_loop;
  }
 
+ /* Add the transaction to the checkpoint list
+  * __journal_remove_checkpoint() can not destroy transaction
+  * under us because it is not marked as T_FINISHED yet */
+ if (journal->j_checkpoint_transactions == NULL) {
+  journal->j_checkpoint_transactions = commit_transaction;
+  commit_transaction->t_cpnext = commit_transaction;
+  commit_transaction->t_cpprev = commit_transaction;
+ } else {
+  commit_transaction->t_cpnext =
+   journal->j_checkpoint_transactions;
+  commit_transaction->t_cpprev =
+   commit_transaction->t_cpnext->t_cpprev;
+  commit_transaction->t_cpnext->t_cpprev =
+   commit_transaction;
+  commit_transaction->t_cpprev->t_cpnext =
+    commit_transaction;
+ }
+ spin_unlock(&journal->j_list_lock);
+
  /* Done with this transaction! */
 
  jbd_debug(3, "JBD2: commit phase 7\n");
@@ -1103,24 +1122,6 @@ restart_loop:
 
  write_unlock(&journal->j_state_lock);
 
- if (journal->j_checkpoint_transactions == NULL) {
-  journal->j_checkpoint_transactions = commit_transaction;
-  commit_transaction->t_cpnext = commit_transaction;
-  commit_transaction->t_cpprev = commit_transaction;
- } else {
-  commit_transaction->t_cpnext =
-   journal->j_checkpoint_transactions;
-  commit_transaction->t_cpprev =
-   commit_transaction->t_cpnext->t_cpprev;
-  commit_transaction->t_cpnext->t_cpprev =
-   commit_transaction;
-  commit_transaction->t_cpprev->t_cpnext =
-    commit_transaction;
- }
- spin_unlock(&journal->j_list_lock);
- /* Drop all spin_locks because commit_callback may be block.
-  * __journal_remove_checkpoint() can not destroy transaction
-  * under us because it is not marked as T_FINISHED yet */
  if (journal->j_commit_callback)
   journal->j_commit_callback(journal, commit_transaction);
 
@@ -1131,7 +1132,7 @@ restart_loop:
  write_lock(&journal->j_state_lock);
  spin_lock(&journal->j_list_lock);
  commit_transaction->t_state = T_FINISHED;
- /* Recheck checkpoint lists after j_list_lock was dropped */
+ /* Check if the transaction can be dropped now that we are finished */
  if (commit_transaction->t_checkpoint_list == NULL &&
      commit_transaction->t_checkpoint_io_list == NULL) {
   __jbd2_journal_drop_transaction(journal, commit_transaction);
-- 
1.7.1